parsec/backend/postgresql/pki.py

```diff
@@ -0,0 +1,439 @@
+    if entry["enrollment_state"] == PkiEnrollmentStatus.SUBMITTED.value:
+        return PkiEnrollmentInfoSubmitted(
+            enrollment_id=entry["enrollment_id"], submitted_on=entry["submitted_on"]
+        )
+    elif entry["enrollment_state"] == PkiEnrollmentStatus.CANCELLED.value:
+        return PkiEnrollmentInfoCancelled(
+            enrollment_id=entry["enrollment_id"],
+            submitted_on=entry["submitted_on"],
+            cancelled_on=entry["info_cancelled"]["cancelled_on"],
+        )
+    elif entry["enrollment_state"] == PkiEnrollmentStatus.REJECTED.value:
+        return PkiEnrollmentInfoRejected(
+            enrollment_id=entry["enrollment_id"],
+            submitted_on=entry["submitted_on"],
+            rejected_on=entry["info_rejected"]["rejected_on"],
+        )
+    elif entry["enrollment_state"] == PkiEnrollmentStatus.ACCEPTED.value:
+        return PkiEnrollmentInfoAccepted(
+            enrollment_id=entry["enrollment_id"],
+            submitted_on=entry["submitted_on"],
+            accepted_on=entry["info_accepted"]["accepted_on"],
+            accepter_der_x509_certificate=entry["info_accepted"]["accepter_der_x509_certificate"],
+            accept_payload_signature=entry["info_accepted"]["accept_payload_signature"],
+            accept_payload=entry["info_accepted"]["accept_payload"],
+        )
+    else:
+        assert False
+
+
+class PGPkiEnrollmentComponent(BasePkiEnrollmentComponent):
+    def __init__(self, dbh: PGHandler):
+        self.dbh = dbh
+
+    def register_components(self, **other_components):
+        self._user_component = other_components["user"]
+
+    async def submit(
+        self,
+        organization_id: OrganizationID,
+        enrollment_id: UUID,
+        force: bool,
+        submitter_der_x509_certificate: bytes,
+        submit_payload_signature: bytes,
+        submit_payload: bytes,
+        submitted_on: DateTime,
+    ) -> None:
+        """
+        Raises:
+            PkiEnrollmentCertificateAlreadySubmittedError
+            PkiEnrollmentAlreadyEnrolledError
+        """
+        submitter_der_x509_certificate_sha1 = hashlib.sha1(submitter_der_x509_certificate).digest()
+        async with self.dbh.pool.acquire() as conn, conn.transaction():
+            # Assert enrollment_id not used
+            row = await conn.fetchrow(
+                *_q_get_pki_enrollment_for_update(
+                    organization_id=organization_id.str, enrollment_id=enrollment_id
+                )
+            )
+            if row:
+                raise PkiEnrollmentIdAlreadyUsedError()
+
+            # Try to retrieve the last attempt with this x509 certificate
+            rep = await conn.fetch(
+                *_q_get_pki_enrollment_from_certificate_sha1(
+                    organization_id=organization_id.str,
+                    submitter_der_x509_certificate_sha1=submitter_der_x509_certificate_sha1,
+                )
+            )
+            for row in rep:
```

Comment:
There is no break in this loop, hence we are checking against all the previous enrollments which seems wrong (we should only access the last enrollment, hence my proposal to rename the query `_q_get_last_pki_enrollment_from_certificate_sha1` and ensure it returns at most 1 result)

Reply:
btw there should a test with enough ACCEPTED/CANCELLED/REJECTED enrollments (and ideally witch a check those old enrollments hasn't been altered) that should break in against this current implementation ;-)

Reply:
and the `_q_get_pki_enrollment_from_certificate_sha1` should do a `FOR UPDATE`

Reply:
There is now a new test that try the following sequence: